pilot/pkg/config/kube/crd/controller/controller.go

```diff
@@ -182,9 +182,25 @@ func (c *controller) createInformer(
 	handler := &kube.ChainHandler{}
 	handler.Append(c.notify)
 
+	// Enhance list function with supplied validation function, so that invalid CRD does not enter store.
+	vlf := func(opts meta_v1.ListOptions) (result runtime.Object, err error) {
+		if result, err = lf(opts); err == nil {
+			// This check is primarily needed for tests which use mock config - but helps otherwise also.
+			if obj, ok := result.(crd.IstioObject); ok {
+				err = vf(obj)
+			}
+		}
+		if err != nil {
+			log.Errorf("failed to add CRD while listing. Value: %v, error: %v", result, err)
+			incrementEvent(otype, "addfailure")
+			return nil, err
+		}
+		return
+	}
+
 	// TODO: finer-grained index (perf)
 	informer := cache.NewSharedIndexInformer(
-		&cache.ListWatch{ListFunc: lf, WatchFunc: wf}, o,
+		&cache.ListWatch{ListFunc: vlf, WatchFunc: wf}, o,
 		resyncPeriod, cache.Indexers{})
```

Comment:
Dumb question: is this going to be invoked each time we construct the push context (where we list a bunch of resources)? If so that would be super costly and redundant for every push context. If not, then 👍

Reply:
It does not. This is invoked only from k8s->store path. Not every time push context is created.